### ISSUE TITLE
Serialise and deserialise don't round-trip for some types (sort of)

### ISSUE THREAD

**Issue by colehaus** [patrick-kidger/equinox#486]
Suppose you have a pytree with a scalar parameter (e.g. `p` on `Dropout`). Now suppose this scalar is a non-standard type like `jax.dtypes.bfloat16`. When you go to serialize this pytree, it's probably already been JAXified (i.e. the scalar is now something like `bf16[]`) (see Fn1). So you'll hit the:

```py
    if isinstance(x, jax.Array):
        jnp.save(f, x)
```
case in `default_serialise_filter_spec` and the saving will work fine. However, when you're deserializing, the `like` pytree will probably still have a scalar (e.g. a raw `bfloat16`). Since `bfloat16` isn't an instance of `float`, you'll hit the final case in `default_deserialise_filter_spec`:

```py
    if isinstance(x, jax.Array):
        return jnp.load(f)
    elif isinstance(x, np.ndarray):
        return np.load(f)
    elif isinstance(x, (bool, float, complex, int)):
        return np.load(f).item()
    else:
        return x
```
But that means the bytes corresponding to the bfloat16 scalar array are still unread on disk and the next `np.load` or `jnp.load` will read from them instead of reading their correct array. Now, the rest of the reads will be off by one until/unless some error occurs later in the deserialization (in my case, calling `item()` on an array with a size greater than zero).

This is fixable in user code. I just ended up doing:

```py
def deserialise_filter_spec(f: BinaryIO, x: A) -> A:
    if isinstance(x, jax.dtypes.bfloat16):
        return jax.dtypes.bfloat16(jnp.load(f).item())
    else:
        return eqx.default_deserialise_filter_spec(f, x)
```

But figuring out what's going on is sort of involved.

Possible improvements:

- Add [`np.generic`](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.generic) to the `isinstance` check in the default spec: `isinstance(x, (bool, float, complex, int, np.generic))`
- Do the `_assert_same` inline with the deserialising instead of afterward to help localize errors. In my case, that means I would've gotten an assertion error in the leaf immediately after the erroneous bfloat16 leaf instead of a deserialization error many leaves later.
- I ended up having to fiddle with `_ordered_tree_map` to print out paths to the leaves to figure out what was going on. It might be nice to include the path in the error message on a failed deserialization.

Though none of these address the root trickiness which is that we generally (I think) want to serialize JAXified pytrees and deserialise based on unJAXified pytrees. `item()` is often sufficient to bridge this gap but not always, as we see here.

Fn1: When I say JAXified, I'm referring to this behavior where scalars that have been passed through a JITed function come back as arrays with zero dimensions. Not sure if there's a proper term for this.

```py
> jax.jit(lambda x: x)(0.1)
Array(0.1, dtype=float32, weak_type=True)
```

(I should be able to look at the weight sharing stuff this weekend. Thanks!)

**Comment by patrick-kidger** [comment]
Yup, we basically just have to trust that `like` is accurate. This is symptomatic of a broader problem in Python, which is that it's very hard to serialise and deserialise reliably.

I think I like all of your suggestion as a way to make this more reliable. I'd be happy to take a PR on any of those!

Regarding JAXification, the general principle in Equinox is to treat these as distinct pytrees -- we never try to treat `array(1)` as `1` or vice versa; these are distinct and unrelated. It's unfortunate that JAX does this implicit promotion; this little detail ends up breaking a lot of things. (Most notably statements like `if foo:`.)

**Comment by patrick-kidger** [comment]
Fixed with #511 and the v0.11.0 release!
